### PR TITLE
SNOW-1926348: Re-implement groupby-apply without pivot

### PR DESCRIPTION
**Pull request by sfc-gh-nkumar**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1926348

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [ ] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#thread-safe-development)
   - [ ] If adding any arguments to public Snowpark APIs or creating new public Snowpark APIs, I acknowledge that I have ensured my changes include AST support. Follow the link for more information: [AST Support Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#ast-abstract-syntax-tree-support-in-snowpark)

3. Please describe how your code solves the related issue.

 This PR re-implements the `groupby-apply` method without the use of pivot. We infer the output schema and use that to create the UDTF with right number of columns. If we fail to infer to schema we fallback to old implementation.
 
**Performance improvement numbers:**
 microbenchmark dataframe_groupby_apply_transform Before: 64.9 sec After: 21.4 sec
 microbenchmark dataframe_groupby_apply_concat Before: 106.2 sec After: 40.7 sec
 microbenchmark dataframe_groupby_apply_agg Before: 14.0 sec After: 13.4 sec
 
 Customer support ticket [Believe](https://snowflakecomputing.atlassian.net/browse/SNOW-1947974):
 [Before](https://app.snowflake.com/sfengineering/sfctest0_aws_us_west_2/#/compute/history/queries?user=NAMED%3ASNOWPANDAS_TEST&start=1743145200000&columns=sqlText%2CqueryId%2Cstatus%2Cuser%2Cwarehouse%2CtotalDuration%2CstartTime&preset=PRESET_LAST_3_DAYS&type=relative&relative=%7B%22tense%22%3A%22past%22%2C%22value%22%3A3%2C%22unit%22%3A%22day%22%2C%22excludePartial%22%3Afalse%2C%22exclusionSize%22%3A%22day%22%2C%22exclusionSizeParam%22%3A%22%22%7D&startDate=2025-03-30T00%3A00%3A00Z&endDate=2025-03-31T00%3A00%3A00Z&page=history%2Fqueries&session_id=3563545552846631):  6 min 21 secs
 [After](https://app.snowflake.com/sfengineering/sfctest0_aws_us_west_2/#/compute/history/queries?user=NAMED%3ASNOWPANDAS_TEST&start=1743145200000&columns=sqlText%2CqueryId%2Cstatus%2Cuser%2Cwarehouse%2CtotalDuration%2CstartTime&preset=PRESET_LAST_3_DAYS&type=relative&relative=%7B%22tense%22%3A%22past%22%2C%22value%22%3A3%2C%22unit%22%3A%22day%22%2C%22excludePartial%22%3Afalse%2C%22exclusionSize%22%3A%22day%22%2C%22exclusionSizeParam%22%3A%22%22%7D&startDate=2025-03-30T00%3A00%3A00Z&endDate=2025-03-31T00%3A00%3A00Z&page=history%2Fqueries&session_id=3563545552865851): 2 min 19 secs
 
 To reduce the scope following will be done in separate PRs:
 we don't infer the columns types yet, that will be a follow up PR.
 We still use old implementation for DataFrame.apply. In a follow up PR I will use new implementation for this as well.

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 1]
```diff
@@ -1874,3 +2042,112 @@ def make_result(value: Any) -> SnowparkColumn:
...
+    # Output ordered dataframe has the following columns in order.
+    # 1. min row position. This is used to find order of group when sort=False.
+    min_row_position_id = output_cols[0][1]
+    # 2. Original row position or row position within group.
+    row_position_id = output_cols[1][1]
+    # 4. index columns (will always include by columns)
+    index_cols = output_cols[2 : 2 + num_index_cols]
+    by_ids = [output_col[1] for output_col in output_cols[2 : 2 + num_by]]
+    # 5. data columns
+    data_cols = output_cols[2 + num_index_cols :]
```
Comment (sfc-gh-jjiao) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:2083]: the comments/descriptions is ranging from 1, 2, to 4, 5?  should it be 1, 2,3,4 ?

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 2]
```diff
@@ -476,10 +526,10 @@ def apply_groupby_func_to_df(
 
     Returns:
         A Tuple of
-         1. rows positions
-         2. Result of applying the function to input dataframe.
-         3. Whether final result should include index columns.
-         4. Whether the index of the result is the same as the index of the input.
+         1. Result of applying the function to input dataframe.
+         2. rows positions
```
Comment (sfc-gh-jjiao) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:530]: qq: do we have to swap the position of the two elements of the returning tuple (  rows positions <-> result of applying the funciton to input dataframe)?  will this make it introduce extra work to change where this function is being called?

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py (hunk 1 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:2083]: good catch.

File: tests/integ/modin/groupby/test_groupby_apply.py [hunk 3]
```diff
@@ -632,9 +624,8 @@ def test_numpy_ints_in_result(
         )
 
     @pytest.mark.xfail(
-        strict=True,
-        raises=NotImplementedError,
-        reason="No support for applying a function that returns two dataframes that have different labels for the column at a given position",
+        raises=AssertionError,
+        reason="If function that returns different labels for different groups, we pick the labels based on dummy input.",
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_apply.py:628]: It feels like this will be surprising to the user?

File: tests/integ/modin/groupby/test_groupby_apply.py [hunk 4]
```diff
@@ -649,10 +640,10 @@ def test_mismatched_data_column_positions(
             ),
         )
 
-    @pytest.mark.xfail(
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_apply.py:652]: Nice!

File: tests/integ/modin/groupby/test_groupby_apply.py [hunk 5]
```diff
@@ -1052,9 +1043,13 @@ def test_dropna(self, dropna, index, include_groups):
             ),
         )
 
-    @pytest.mark.xfail(raises=NotImplementedError, strict=True, reason="SNOW-1232201")
+    @pytest.mark.xfail(
+        raises=AssertionError,
+        strict=True,
+        reason="Snowpark pandas uses name returned from first group, while pandas returns None",
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_apply.py:1049]: Can't we return None as well?

File: tests/integ/modin/groupby/test_groupby_apply.py [hunk 6]
```diff
@@ -1075,9 +1070,13 @@ def test_returning_series_with_different_names(
             ),
         )
 
-    @pytest.mark.xfail(raises=NotImplementedError, strict=True, reason="SNOW-1232208")
+    @pytest.mark.xfail(
+        raises=AssertionError,
+        strict=True,
+        reason="Snowpark pandas return a DataFrame but native pandas returns a Series",
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_apply.py:1076]: Would NotImplementedError be better here until we can be more compatible with pandas? Returning DataFrame instead of Series seems like a big difference.

File: tests/integ/modin/groupby/test_groupby_apply.py [hunk 7]
```diff
@@ -1257,8 +1251,8 @@ class TestCallableWithMixedReturnTypes:
 
     @pytest.mark.xfail(
         strict=True,
-        raises=NotImplementedError,
-        reason="NotImplementedError in Snowpark pandas. see SNOW-1236959",
+        raises=SnowparkSQLException,
+        reason="Results in mismatch between udtf schema and the actual result.",
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_apply.py:1255]: How significant is the mismatch? Would it be tolerable by the user?

File: tests/integ/modin/groupby/test_groupby_apply.py [hunk 8]
```diff
@@ -1273,7 +1267,7 @@ def test_scalar_then_dataframe(self):
     @pytest.mark.xfail(
         strict=True,
         raises=AssertionError,
-        reason="Snowpark pandas transposes the series, but pandas doesn't. See SNOW-1236959",
+        reason="Results in mismatch between udtf schema and the actual result.",
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_apply.py:1270]: Ditto.

File: tests/integ/modin/groupby/test_groupby_apply.py [hunk 9]
```diff
@@ -160,8 +160,7 @@ def grouping_dfs_with_multiindexes() -> tuple[pd.DataFrame, native_pd.DataFrame]
 # For cases where we need to check whether we have a transform, we add an extra
 # query between 4) and 5) to check whether the function acted as a transform.
 
-QUERY_COUNT_WITHOUT_TRANSFORM_CHECK = 5
-QUERY_COUNT_WITH_TRANSFORM_CHECK = 6
+QUERY_COUNT = 4
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_apply.py:163]: At a high level, where is the saving in query count coming from after avoiding the pivot step?

File: src/snowflake/snowpark/modin/plugin/docstrings/groupby.py [hunk 10]
```diff
@@ -1122,20 +1122,20 @@ def apply():
         each group together into a new DataFrame:
 
         >>> g1[['B', 'C']].apply(lambda x: x.select_dtypes('number') / x.select_dtypes('number').sum())
-                    B    C
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/docstrings/groupby.py:1125]: Curious why did these need to change?

File: tests/integ/modin/groupby/test_groupby_apply.py (hunk 9 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/groupby/test_groupby_apply.py:163]: we had forced materialization in old implementation https://github.com/snowflakedb/snowpark-python/blob/03dab33abcdfb2b6d084abe5fc985e1b03f42d0e/src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py#L4245

Another query count reduction (in some cases) is from avoiding transform check. In new implementation, we don't have to do transform check query as we infer it using dummy data.

File: tests/integ/modin/groupby/test_groupby_apply.py (hunk 3 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/groupby/test_groupby_apply.py:628]: This is the edge case where use can potentially write a apply function which will return different schema for different groups. Unfortunately we don't have good solution for it but I wound't worry about it as this should be very rare.

File: tests/integ/modin/groupby/test_groupby_apply.py (hunk 5 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/groupby/test_groupby_apply.py:1049]: No, we rely on dummy input to infer schema. Only way to know for sure is run it on actual data, that would be cost prohibitive.

File: tests/integ/modin/groupby/test_groupby_apply.py (hunk 6 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/groupby/test_groupby_apply.py:1076]: No way  to detect unless we infer the schema using full actual data which is again very expensive. But on high level these cases where apply function returns different schema for different groups should be very rare.
We did similar [analysis](https://docs.google.com/document/d/1xaxDVqjYn_cmlb4pUNqr1vRb1SXhDePAP74Kdhxzfo8/edit?tab=t.0#heading=h.6bjjpt145epq) for df.applymap and series.apply methods and found that users usually don't write such functions where schema varies based on input data. If we extend the same intuition for groupby.apply this should also be very rare.

File: tests/integ/modin/groupby/test_groupby_apply.py (hunk 7 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/groupby/test_groupby_apply.py:1255]: This the edge case where apply function returns scalar for one group and dataframe for another group. I wouldn't worry about this.

File: tests/integ/modin/groupby/test_groupby_apply.py (hunk 6 above)
Comment (sfc-gh-helmeleegy) [reply tests/integ/modin/groupby/test_groupby_apply.py:1076]: That makes sense. I think it's worth adding a comment along with the tests explaining that we anticipate these cases to be very rare.

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM, thanks Naresh!

File: tests/integ/modin/groupby/test_groupby_apply.py (hunk 9 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/groupby/test_groupby_apply.py:163]: Yeah, it was there to make pivot more efficient. we don't this anymore.

File: tests/integ/modin/groupby/test_groupby_apply.py (hunk 6 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/groupby/test_groupby_apply.py]: Added comments to tests.

**Review by sfc-gh-jjiao — APPROVED** [review]

**Review by sfc-gh-mvashishtha — CHANGES_REQUESTED** [review]
@sfc-gh-nkumar, @sfc-gh-jjiao asked me to review if I have time, so please hold till I review. ETA for my review is in 15 minutes. Thank you!

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 11]
```diff
@@ -1874,3 +2042,112 @@ def make_result(value: Any) -> SnowparkColumn:
...
+    ordered_dataframe: OrderedDataFrame,
+    output_schema: GroupbyApplyOutputSchema,
+    num_by: int,
+    is_transform: bool,
+    group_keys: bool,
+    as_index: bool,
+    sort: bool,
+) -> InternalFrame:
+    """
+    Create Interframe for groupy apply no pivot output.
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:2058]:
```suggestion
    Create InternalFrame for groupy apply no pivot output.
```

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 12]
```diff
@@ -1874,3 +2042,112 @@ def make_result(value: Any) -> SnowparkColumn:
...
+    Args:
+        input_frame: Input internal frame.
+        ordered_dataframe: Output ordered dataframe from groupby apply udtf.
+        output_schema: Inferred output schema for groupby apply.
+        num_by: Number of by columns.
+        is_transform: Weather this is groupy.transform or not.
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:2064]:
```suggestion
        is_transform: Whether this is groupy.transform or not.
```

File: tests/integ/modin/groupby/test_groupby_apply.py [hunk 13]
```diff
@@ -154,14 +154,12 @@ def grouping_dfs_with_multiindexes() -> tuple[pd.DataFrame, native_pd.DataFrame]
 # 2. Put a zip file with the UDTF definition in the temporary stage (we filter this out when counting queries)
 # 3. Create the UDTF (increasing UDTF count by 1)
 # 3. Apply the UDTF using a join (increasing join count by 1)
-# 4. Save UDTF result as temp table to work around SNOW-1060191
-# 5. convert result to pandas
+# 4. convert result to pandas
```
Comment (sfc-gh-mvashishtha) [inline tests/integ/modin/groupby/test_groupby_apply.py:157]: We should fix the description of the query count, e.g. it says "For almost all test cases, the query count is either 5 or 6,"

File: tests/integ/modin/groupby/test_groupby_apply.py [hunk 14]
```diff
@@ -1075,9 +1072,14 @@ def test_returning_series_with_different_names(
             ),
         )
 
-    @pytest.mark.xfail(raises=NotImplementedError, strict=True, reason="SNOW-1232208")
+    @pytest.mark.xfail(
+        raises=AssertionError,
+        strict=True,
+        reason="Snowpark pandas return a DataFrame but native pandas returns a"
+        + " Series. This should be very rate in practice.",
```
Comment (sfc-gh-mvashishtha) [inline tests/integ/modin/groupby/test_groupby_apply.py:1079]:
```suggestion
        + " Series. This should be very rare in practice.",
```

**Review by sfc-gh-mvashishtha — APPROVED** [review]
Wow, the groupby.apply code is complicated. Nice job, though! I'm leaving minor comments.
